Allow admin/user to filter by state for Pods, Clusters and Hosts

kishankavala opened the pull request:
### Description

This PR allows admin/user to filter by state for Pods, Clusters and Hosts. This is part of https://github.com/apache/cloudstack/issues/7366 .

Filters available on UI : `Enabled`, `Disabled`

Pods, Clusters: Filter by allocation state
Hosts: Filter by resource state


### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial


### Screenshots (if appropriate):


### How Has This Been Tested?
Tested UI changes manually


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

rohityadavcloud commented:
@blueorangutan package

rohityadavcloud commented on ui/src/config/section/infra/hosts.js:
```diff
@@ -24,6 +24,10 @@ export default {
   permission: ['listHostsMetrics'],
   resourceType: 'Host',
+  filters: () => {
+    const filters = ['enabled', 'disabled']
```
@kishankavala could you explore other states for the resources. For example as an admin I only want to see hosts in Alert, Maintenance and other useful states? Pl do the same for other resources (pods, cluster) in this PR

shwstppr replied:
@rohityadavcloud cc @kishankavala I guess this PR adds filter for host `resourcestate` which can only be,
```
Creating,
Enabled,
Disabled,
ErrorInPrepareForMaintenance,
PrepareForMaintenance,
ErrorInMaintenance,
Maintenance,
Error,
Degraded
```

Not sure if we have filtering for host state as well which can be host `status` as well which can be,

```
Creating
Connecting
Up
Down
Disconnected
Alert
Removed
Error
Rebalancing
Unknown
```

rohityadavcloud replied:
@kishankavala any update on this? Thanks.

rohityadavcloud commented:
cc @vishesh92

vishesh92 commented:
Checking this PR in a bit

rohityadavcloud commented:
@blueorangutan ui

blueorangutan commented:
@rohityadavcloud a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: https://qa.cloudstack.cloud/simulator/pr/7373 (QA-JID-66)

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rohityadavcloud reviewed (changes requested):
Overall LGTM. Tested UI changes, I think the host state list could be added to have all the possible states for hosts here, which may be useful for the operator such as Alert etc. Also for host, the agent/state may be a more useful state to filter on than the resource state. The agent/host status can be Up/Down/Alert... see https://github.com/apache/cloudstack/blob/main/api/src/main/java/com/cloud/host/Status.java#L26

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 5841

rohityadavcloud commented:
We're pending one change, otherwise this can be merged. @kishankavala could you address the review comments, thanks.

kishankavala commented:
Making changes to add suggested host stated

kishankavala commented:
Cluster has AllocationState and ManagedState. Filter uses AllocationState. ManagedState  can be added if required
 - AllocationState {
        Enabled, Disabled;
    };
- ManagedState {
        Managed, PrepareUnmanaged, Unmanaged, PrepareUnmanagedError
    };

kishankavala commented:
Pod has only allocation state and is added to the filter.
- AllocationState {
        Enabled, Disabled;
    };

kishankavala commented:
Host has State and ResourceState. Its a long list even after removing transition states. 
- State {
    Up,
    Down,
    Disconnected,
    Alert,
    Removed,
    Error,
    Unknown
}

- ResourceState {
    Enabled,
    Disabled,
    ErrorInPrepareForMaintenance,
    PrepareForMaintenance,
    ErrorInMaintenance,
    Maintenance,
    Error,
    Degraded
}

Added the Up, Down, Alert, Enabled, Disabled, Maintenance in the filter

kishankavala commented:
@blueorangutan ui

blueorangutan commented:
@kishankavala a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: https://qa.cloudstack.cloud/simulator/pr/7373 (QA-JID-87)

rohityadavcloud commented on ui/src/views/AutogenView.vue:
```diff
@@ -1573,6 +1575,20 @@ export default {
+        }
+      } else if (['host'].includes(this.$route.name)) {
+        if (filter === 'all') {
+          delete query.resourcestate
+        } else if (['up', 'down', 'alert'].includes(filter)) {
+          query.state = filter
+        } else {
+          query.resourcestate = filter
```
```suggestion
          delete query.resourcestate
          delete query.state
        } else if (['up', 'down', 'alert'].includes(filter)) {
          delete query.resourcestate
          query.state = filter
        } else {
          delete query.state
          query.resourcestate = filter
```

rohityadavcloud commented:
@blueorangutan ui

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: https://qa.cloudstack.cloud/simulator/pr/7373 (QA-JID-89)

shwstppr reviewed (approved):
LGTM. Testing filtering resources on qa server
